### PR TITLE
.github: Migrate ecr_gc to github actions

### LINKED ISSUE #62269: `docker_for_ecr_gc_build_job` broken on master CI due to reaching maximum number of images on Amazon ECR
## 🐛 Bug

CI test (`docker_for_ecr_gc_build_job`) on Master is consistently failing:
When `push refers to repository [308535385114.dkr.ecr.us-east-1.amazonaws.com/gc/ecr]`, it raises `denied: The repository with name 'gc/ecr' in registry with id '308535385114' already has the maximum allowed number of images which is '10000'`

Example test:
https://app.circleci.com/pipelines/github/pytorch/pytorch/356564/workflows/3a3aa850-19eb-4d86-900a-f2f5e8a228c6/jobs/15031783

## Task list (from @seemethere)

- [x] Remove intermediate build images with an ECR lifecycle policy (https://github.com/pytorch/pytorch/pull/62610)
- [ ] Remove preview base docker image builds with an ECR lifecycle policy

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @seemethere @malfet @walterddr @pytorch/pytorch-dev-infra

### PR DESCRIPTION
**Pull request by seemethere**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62286
* __->__ #62284

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29943356](https://our.internmc.facebook.com/intern/diff/D29943356)

### CODE REVIEW

**Comment by seemethere** [comment]
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29943356).

File: .github/workflows/ecr_gc_docker.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+on:
+  schedule:
+    # Runs every 30 minutes
+    - cron: "*/30 * * * *"
+  # Have the ability to trigger this job manually using the API as well
+  workflow_dispatch:
+
+env:
+  AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ECR_GC_ACCESS_KEY_ID }}
```
Comment (seemethere) [inline .github/workflows/ecr_gc_docker.yml:11]: I've already went through and added these secrets here

File: .circleci/ecr_gc_docker/gc.py [hunk 2]
```diff
@@ -7,6 +7,69 @@
 import re
 import sys
 
+# TODO: See if this is actually still needed
+# Default list of legacy tags to keep
+IGNORE_TAGS = {
```
Comment (seemethere) [inline .circleci/ecr_gc_docker/gc.py:12]: I'm not entirely sure if we actually need these anymore

File: .github/workflows/ecr_gc_docker.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      - name: Generate tags to keep
+        run: |
+          GENERATED_IMAGE_TAG=$(\
+            git log --oneline --pretty='%H' .circleci/docker \
+              | xargs -I '{}' git rev-parse '{}:.circleci/docker' \
+              | paste -sd "," -)
+          echo "export GENERATED_IMAGE_TAG='${GENERATED_IMAGE_TAG}'" >> "${BASH_ENV}"
+      - name: Run garbage collection
+        run: |
+          .circleci/ecr_gc_docker/gc.py --filter-prefix "${FILTER_PREFIX}" --ignore-tags "${GENERATED_IMAGE_TAG}"
```
Comment (zhouzhuojie) [inline .github/workflows/ecr_gc_docker.yml:43]: oh nice, get rid of the gc image all together! 🎉

**Review by zhouzhuojie — APPROVED** [review]

**Review by malfet — APPROVED** [review]
LGTM, though I wonder if we shoudl also delete `.circleci/verbatim-sources/workflows/workflows-ecr-gc.yml` as well part of `.circleci/verbatim-sources/job-specs/docker_jobs.yml`

**Comment by seemethere** [comment]
> LGTM, though I wonder if we shoudl also delete `.circleci/verbatim-sources/workflows/workflows-ecr-gc.yml` as well part of `.circleci/verbatim-sources/job-specs/docker_jobs.yml`

We can delete them in a follow up after we verify the workflow actually works

File: .github/workflows/ecr_gc_docker.yml [hunk 4]
```diff
@@ -0,0 +1,43 @@
+name: Garbage Collect ECR Images
+
+on:
+  schedule:
+    # Runs every 30 minutes
+    - cron: "*/30 * * * *"
+  # Have the ability to trigger this job manually using the API as well
+  workflow_dispatch:
```
Comment (zhouzhuojie) [inline .github/workflows/ecr_gc_docker.yml:8]: is it green when you manually triggering it?

Comment (seemethere) [reply .github/workflows/ecr_gc_docker.yml:8]: Can't actually manually trigger it until we merge first unfortunately.

Would be a good use for a canary repository though

**Comment by seemethere** [comment]
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29943356).

**Review by walterddr — APPROVED** [review]

File: .github/workflows/ecr_gc_docker.yml (hunk 4 above)
Comment (zhouzhuojie) [reply .github/workflows/ecr_gc_docker.yml:8]: I think we can add on pull_request temporally to test the script, and then before the merge remove the on pull_request trigger

**Pull request closed without merging**